File: turnier-app/frontend/src/app/services/turnier.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Mannschaft } from '../models/mannschaft.model';
import { Gruppe } from '../models/gruppe.model';
import { Spiel } from '../models/spiel.model';
import { TurnierStatus } from '../models/turnier-status.model';

@Injectable({
  providedIn: 'root'
})
export class TurnierService {
  private apiUrl = 'http://localhost:8081/api/turnier';

  constructor(private http: HttpClient) { }

  // Mannschaften
  createMannschaft(name: string): Observable<Mannschaft> {
    return this.http.post<Mannschaft>(`${this.apiUrl}/mannschaften`, { name });
  }

  getAllMannschaften(): Observable<Mannschaft[]> {
    return this.http.get<Mannschaft[]>(`${this.apiUrl}/mannschaften`);
  }

  deleteMannschaft(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/mannschaften/${id}`);
  }

  // Gruppen
  erstelleGruppen(): Observable<any> {
    return this.http.post(`${this.apiUrl}/gruppen/erstellen`, {});
  }

  getAllGruppen(): Observable<Gruppe[]> {
    return this.http.get<Gruppe[]>(`${this.apiUrl}/gruppen`);
  }

  getGruppentabelle(gruppenId: number): Observable<Mannschaft[]> {
    return this.http.get<Mannschaft[]>(`${this.apiUrl}/gruppen/${gruppenId}/tabelle`);
  }

  getGruppenspieleById(gruppenId: number): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/gruppen/${gruppenId}/spiele`);
  }

  erstelleGruppenspiele(): Observable<any> {
    return this.http.post(`${this.apiUrl}/gruppenspiele/erstellen`, {});
  }

  // Spiele
  getGruppenspiele(): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/spiele/gruppenspiele`);
  }

  getViertelfinale(): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/spiele/viertelfinale`);
  }

  getHalbfinale(): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/spiele/halbfinale`);
  }

  getFinale(): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/spiele/finale`);
  }

  getAktuelleRunde(): Observable<Spiel[]> {
    return this.http.get<Spiel[]>(`${this.apiUrl}/spiele/aktuelle-runde`);
  }

  updateSpielErgebnis(spielId: number, tore1: number, tore2: number): Observable<Spiel> {
    return this.http.put<Spiel>(`${this.apiUrl}/spiele/${spielId}/ergebnis`, { tore1, tore2 });
  }

  // Turnier
  erstelleNaechsteRunde(): Observable<any> {
    return this.http.post(`${this.apiUrl}/naechste-runde`, {});
  }

  getTurnierStatus(): Observable<TurnierStatus> {
    return this.http.get<TurnierStatus>(`${this.apiUrl}/status`);
  }

  resetTurnier(): Observable<any> {
    return this.http.post(`${this.apiUrl}/reset`, {});
  }
}

File: turnier-app/src/main/java/com/turnier/service/TurnierService.java
package com.turnier.service;

import com.turnier.entity.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
@Transactional
public class TurnierService {
    
    @Autowired
    private MannschaftService mannschaftService;
    
    @Autowired
    private GruppeService gruppeService;
    
    @Autowired
    private SpielService spielService;
    
    @Autowired
    private StatistikenService statistikenService;
    
    @Autowired
    private TurnierPhaseService turnierPhaseService;
    
    // Delegierungsmethoden für MannschaftService
    public Mannschaft createMannschaft(String name) {
        return mannschaftService.createMannschaft(name);
    }
    
    public List<Mannschaft> getAllMannschaften() {
        return mannschaftService.getAllMannschaften();
    }
    
    public void deleteMannschaft(Long id) {
        mannschaftService.deleteMannschaft(id);
    }
    
    // Delegierungsmethoden für GruppeService
    public void erstelleGruppen() {
        long anzahlMannschaften = mannschaftService.getAnzahlMannschaften();
        gruppeService.erstelleGruppen(anzahlMannschaften);
    }
    
    public List<Gruppe> getAllGruppen() {
        return gruppeService.getAllGruppen();
    }
    
    public List<Mannschaft> getGruppentabelle(Long gruppenId) {
        return gruppeService.getGruppentabelle(gruppenId);
    }
    
    public List<Spiel> getGruppenspiele(Long gruppenId) {
        return gruppeService.getGruppenspiele(gruppenId);
    }
    
    // Delegierungsmethoden für SpielService
    public List<Spiel> getSpieleByTyp(Spiel.SpielTyp spielTyp) {
        return spielService.getSpieleByTyp(spielTyp);
    }
    
    public List<Spiel> getAktuelleRunde() {
        return spielService.getAktuelleRunde();
    }
    
    public Spiel updateSpielErgebnis(Long spielId, Integer tore1, Integer tore2) {
        Spiel spiel = spielService.updateSpielErgebnis(spielId, tore1, tore2);
        
        // Aktualisiere Mannschaftsstatistiken
        statistikenService.aktualisiereMannschaftsstatistiken(spiel);
        
        // Für Gruppenspiele: Prüfe ob Gruppe abgeschlossen ist
        if (spiel.getSpielTyp() == Spiel.SpielTyp.GRUPPENSPIEL) {
            gruppeService.prüfeGruppenphaseAbgeschlossen();
        }
        
        return spiel;
    }
    
    public Mannschaft getGewinner(Spiel spiel) {
        return spielService.getGewinner(spiel);
    }
    
    // Delegierungsmethoden für TurnierPhaseService
    public void erstelleGruppenspiele() {
        turnierPhaseService.erstelleGruppenspiele();
    }
    
    public void erstelleNächsteRunde() {
        turnierPhaseService.erstelleNächsteRunde();
    }
    
    public void starteKOPHase() {
        turnierPhaseService.starteKOPHase();
    }
    
    public Map<String, Object> getTurnierStatus() {
        Map<String, Object> status = new HashMap<>();
        
        // Hole Finale
        List<Spiel> finale = spielService.getSpieleByTyp(Spiel.SpielTyp.FINALE);
        if (!finale.isEmpty() && finale.get(0).isIstBeendet()) {
            Mannschaft sieger = spielService.getGewinner(finale.get(0));
            status.put("sieger", sieger.getName());
            status.put("istBeendet", true);
        } else {
            status.put("sieger", null);
            status.put("istBeendet", false);
        }
        
        return status;
    }
}
